File: plc4j/spi/src/main/java/org/apache/plc4x/java/spi/codegen/io/DataReaderFactory.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.plc4x.java.spi.codegen.io;

import org.apache.plc4x.java.spi.generation.ReadBuffer;

import java.math.BigInteger;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.function.Function;

public class DataReaderFactory {

    public static DataReader<Boolean> readBoolean(ReadBuffer readBuffer) {
        return new DataReaderSimpleBoolean(readBuffer);
    }

    public static DataReader<Byte> readUnsignedByte(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleUnsignedByte(readBuffer, bitLength);
    }

    public static DataReader<Byte> readByte(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleByte(readBuffer, bitLength);
    }

    public static DataReader<Short> readUnsignedShort(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleUnsignedShort(readBuffer, bitLength);
    }

    public static DataReader<Integer> readUnsignedInt(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleUnsignedInt(readBuffer, bitLength);
    }

    public static DataReader<Long> readUnsignedLong(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleUnsignedLong(readBuffer, bitLength);
    }

    public static DataReader<BigInteger> readUnsignedBigInteger(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleUnsignedBigInteger(readBuffer, bitLength);
    }

    public static DataReader<Byte> readSignedByte(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleSignedByte(readBuffer, bitLength);
    }

    public static DataReader<Short> readSignedShort(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleSignedShort(readBuffer, bitLength);
    }

    public static DataReader<Integer> readSignedInt(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleSignedInt(readBuffer, bitLength);
    }

    public static DataReader<Long> readSignedLong(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleSignedLong(readBuffer, bitLength);
    }

    public static DataReader<BigInteger> readSignedBigInteger(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleSignedBigInteger(readBuffer, bitLength);
    }

    public static DataReader<Float> readFloat(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleFloat(readBuffer, bitLength);
    }

    public static DataReader<Double> readDouble(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleDouble(readBuffer, bitLength);
    }

    public static DataReader<String> readString(ReadBuffer readBuffer, int bitLength) {
        return new DataReaderSimpleString(readBuffer, bitLength);
    }

    public static <T, I> DataReaderEnumDefault<T, I> readEnum(Function<I, T> enumResolver, DataReader<I> dataReader) {
        return new DataReaderEnumDefault<>(enumResolver, dataReader);
    }

    public static DataReader<LocalDate> readDate(ReadBuffer readBuffer) {
        return new DataReaderSimpleDate(readBuffer);
    }

    public static DataReader<LocalDateTime> readDateTime(ReadBuffer readBuffer) {
        return new DataReaderSimpleDateTime(readBuffer);
    }

    public static DataReader<LocalTime> readTime(ReadBuffer readBuffer) {
        return new DataReaderSimpleTime(readBuffer);
    }

}

File: plc4go/spi/codegen/io/DataReaderFactory.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package io

import (
	"context"
	"math/big"
	"time"

	"github.com/apache/plc4x/plc4go/spi/utils"
)

func ReadBoolean(readBuffer utils.ReadBuffer) DataReader[bool] {
	return NewDataReaderSimpleBoolean(readBuffer)
}

func ReadUnsignedByte(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[byte] {
	return NewDataReaderSimpleUnsignedByte(readBuffer, bitLength)
}

func ReadByte(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[byte] {
	return NewDataReaderSimpleByte(readBuffer, bitLength)
}

func ReadUnsignedShort(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[uint16] {
	return NewDataReaderSimpleUnsignedShort(readBuffer, bitLength)
}

func ReadUnsignedInt(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[uint32] {
	return NewDataReaderSimpleUnsignedInt(readBuffer, bitLength)
}

func ReadUnsignedLong(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[uint64] {
	return NewDataReaderSimpleUnsignedLong(readBuffer, bitLength)
}

func ReadUnsignedBigInteger(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[*big.Int] {
	return NewDataReaderSimpleUnsignedBigInteger(readBuffer, bitLength)
}

func ReadSignedByte(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[int8] {
	return NewDataReaderSimpleSignedByte(readBuffer, bitLength)
}

func ReadSignedShort(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[int16] {
	return NewDataReaderSimpleSignedShort(readBuffer, bitLength)
}

func ReadSignedInt(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[int32] {
	return NewDataReaderSimpleSignedInt(readBuffer, bitLength)
}

func ReadSignedLong(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[int64] {
	return NewDataReaderSimpleSignedLong(readBuffer, bitLength)
}

func ReadSignedBigInteger(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[*big.Int] {
	return NewDataReaderSimpleSignedBigInteger(readBuffer, bitLength)
}

func ReadFloat(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[float32] {
	return NewDataReaderSimpleFloat(readBuffer, bitLength)
}

func ReadDouble(readBuffer utils.ReadBuffer, bitLength uint8) DataReader[float64] {
	return NewDataReaderSimpleDouble(readBuffer, bitLength)
}

func ReadString(readBuffer utils.ReadBuffer, bitLength uint32) DataReader[string] {
	return NewDataReaderSimpleString(readBuffer, bitLength)
}

func ReadEnum[T any, I any](enumResolver func(I) (T, bool), dataReader DataReader[I]) *DataReaderEnumDefault[T, I] {
	return NewDataReaderEnumDefault[T, I](enumResolver, dataReader)
}

func ReadComplex[T any](complexSupplier func(context.Context, utils.ReadBuffer) (T, error), readBuffer utils.ReadBuffer) *DataReaderComplexDefault[T] {
	return NewDataReaderComplexDefault[T](complexSupplier, readBuffer)
}

func ReadDate(readBuffer utils.ReadBuffer) DataReader[time.Time] {
	return NewDataReaderSimpleDate(readBuffer)
}

func ReadDateTime(readBuffer utils.ReadBuffer) DataReader[time.Time] {
	return NewDataReaderSimpleDateTime(readBuffer)
}

func ReadTime(readBuffer utils.ReadBuffer) DataReader[time.Time] {
	return NewDataReaderSimpleTime(readBuffer)
}
